tests/Freckle/App/OpenTelemetry/ContextSpec.hs

```diff
@@ -0,0 +1,156 @@
+
+  describe "processWithContext" $ do
+    it "creates a child-span context" $ appExample $ do
+      inSpan "example" defaultSpanArguments $ do
+        outerSpanContext <- assertCurrentSpanContext
+
+        let
+          traceId = traceIdToHex $ Trace.traceId outerSpanContext
+          spanId = spanIdToHex $ Trace.spanId outerSpanContext
+
+          headers :: [Header]
+          headers =
+            [ ("traceparent", encodeUtf8 $ toTraceParent traceId spanId)
+            , ("tracestate", "")
+            ]
+
+        processWithContext "loop" defaultSpanArguments headers $ \_ -> do
+          innerSpanContext <- assertCurrentSpanContext
+
+          -- Ideally we'd assert the inner span's parent is the outer span, but
+          -- the library doesn't expose the necessary functionality. All we can
+          -- do is assert it's a fresh span of the same trace.
+          Trace.traceId innerSpanContext `shouldBe` Trace.traceId outerSpanContext
+          Trace.spanId innerSpanContext `shouldNotBe` Trace.spanId outerSpanContext
+
+    it "sets the current context back into the headers" $ appExample $ do
+      inSpan "example" defaultSpanArguments $ do
+        let
+          traceId = "fba7cd19bff3ef866a599d6f6a85baef"
+          spanId = "5747fd6b144f4009"
+
+          headers :: [Header]
+          headers =
+            [ ("traceparent", encodeUtf8 $ toTraceParent traceId spanId)
+            , ("tracestate", "")
+            ]
+
+        processWithContext "loop" defaultSpanArguments headers $ \headers' -> do
+          spanContext <- assertCurrentSpanContext
+
+          let headerTraceParent = do
+                bs <- lookup "traceparent" headers'
+                fromTraceParent $ decodeUtf8 bs
+
+          fmap fst headerTraceParent `shouldBe` Just traceId
+          fmap snd headerTraceParent
+            `shouldBe` Just (spanIdToHex $ Trace.spanId spanContext)
+
+    it "sets a fresh context back into the headers" $ appExample $ do
+      inSpan "example" defaultSpanArguments $ do
+        let
+          headers :: [Header]
+          headers = []
+
+        processWithContext "loop" defaultSpanArguments headers $ \headers' -> do
+          spanContext <- assertCurrentSpanContext
+
+          let headerTraceParent = do
+                bs <- lookup "traceparent" headers'
+                fromTraceParent $ decodeUtf8 bs
+
+          fmap fst headerTraceParent
+            `shouldBe` Just (traceIdToHex $ Trace.traceId spanContext)
+          fmap snd headerTraceParent
+            `shouldBe` Just (spanIdToHex $ Trace.spanId spanContext)
+
+assertCurrentSpanContext :: (MonadIO m, HasCallStack) => m Trace.SpanContext
+assertCurrentSpanContext =
+  getCurrentSpanContext >>= \case
+    Nothing -> expectationFailure "Expect there to be a SpanContext"
+    Just spanContext -> pure spanContext
+
+toTraceParent :: Text -> Text -> Text
+toTraceParent traceId spanId = "00-" <> traceId <> "-" <> spanId <> "-00"
```

Comment:
What's up with needing the prefix and suffix?

Reply:
This just just how OpenTelemetry formats the `traceparent` header it uses in context propagation. I'm sure we could look up the spec to see what the values mean. I'm sure `00` is just what I'm getting in test because there's no sampling to encode, or something like that.